pow-migration: Allow stakers to increase stake and switch validators

jsdanielh opened the pull request:
- Check that the registered stakers and stakers created from validator deposits that exceed the validator minimum deposit are always greater than the minimum stake. Otherwise keep them as burnt coins.
- Use the total balance used to register stakers and validators to decrease it from the burn address balance in the PoS genesis in order to preserve the total supply from the PoW chain.
- Allow stakers to increase stake and switch validators by ordering the transactions per sender, then keeping track of stakers being added and then increasing the staker and/or updating the delegation if there were previous valid transactions.
  This closes #2352, closes #2353 and closes #2362.
- Sort the possible validator registration transactions by timestamp before parsing them.

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
pow-migration: Sort validator registration transactions

Sort the possible validator registration transactions by timestamp
before parsing them.

jsdanielh committed:
pow-migration: Allow stakers to increase stake and switch validator

Allow stakers to increase stake and switch validators by ordering the
transactions per sender, then keeping track of stakers being added
and then increasing the staker and/or updating the delegation if
there were previous valid transactions.

This closes #2352, closes #2353 and closes #2362.

jsdanielh committed:
pow-migration: Preserve the total supply from the PoW chain

Use the total balance used to register stakers and validators to
decrease it from the burn address balance in the PoS genesis in order
to preserve the total supply from the PoW chain.

sisou commented on pow-migration/src/state/mod.rs:
```diff
@@ -315,13 +325,18 @@ pub async fn get_stakers(
-                inactive_balance: Coin::ZERO,
-                inactive_from: None,
-            })
+            let staker_address = validator.validator.validator_address.clone();
```
```suggestion
            let delegation_address = validator.validator.validator_address.clone();
```

sisou replied:
Having a struct `validator` inside a struct `validator` seems like bad naming.

sisou commented on pow-migration/src/state/mod.rs:
```diff
@@ -315,13 +325,18 @@ pub async fn get_stakers(
+                    balance: validator.balance
+                        - Coin::from_u64_unchecked(Policy::VALIDATOR_DEPOSIT),
+                    delegation: validator.validator.validator_address.clone(),
+                    inactive_balance: Coin::ZERO,
+                    inactive_from: None,
```
Why even have the fields `inactive_balance` and `inactive_from` in `GenesisStaker`? Aren't they always ZERO and None? You also don't have the `retired_*` fields.

jsdanielh replied:
These are fields required for building the genesis. In our case they are always zero but they come from the genesis crate which exposes them. The retired fields are not part of a staker definition.

sisou commented on pow-migration/src/state/mod.rs:
```diff
@@ -337,7 +352,9 @@ pub async fn get_stakers(
     }
 
     // Now look for the commit transaction
```
This comment seems wrong. It's not looking for the validator commit transaction here, but for pre-stake transactions.

sisou reviewed (commented):
Does this script check for the minimum stake when creating a staker? I think it should, and stakers below the threshold must be skipped and burnt.

Eligioo commented on pow-migration/src/genesis/mod.rs:
```diff
@@ -110,6 +109,21 @@ pub async fn get_pos_genesis(
+    let mut burnt_registration_balance = genesis_stakers.iter().map(|staker| staker.balance).sum();
+    burnt_registration_balance += genesis_validators
+        .iter()
+        .map(|validator| validator.balance)
```
Using the `fold` combinator here is the equivalent of `map().sum()` without mapping the struct first.

jsdanielh committed:
pow-migration: Check for minimum stake

Check that the registered stakers and stakers created from validator
deposits that exceed the validator minimum deposit are always greater
than the minimum stake. Otherwise keep them as burnt coins.

jsdanielh commented:
> Does this script check for the minimum stake when creating a staker? I think it should, and stakers below the threshold must be skipped and burnt.

You're right. Added this check as well

viquezclaudio reviewed (approved):
LGTM
